kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java

```diff
@@ -0,0 +1,319 @@
+                                                        restMapSet
+                                                ),
+                                                BooleanUtils.FALSE
+                                        ),
+                                        setArgsCeil
+                                )
+                        )
+                        , BooleanUtils.TRUE
+                        , BooleanUtils.TRUE
+                        , Att.empty().add(Att.SIMPLIFICATION())
+                );
+        rules.add(ceilMapRule);
+    }
+
+    private boolean isRealHook(Att att) {
+        String hook = att.get(Att.HOOK());
+        if (hook.startsWith("ARRAY.")) {
+            return false;
+        }
+        if (options.hookNamespaces.stream().anyMatch(ns -> hook.startsWith(ns + "."))) {
+            return true;
+        }
+        return Hooks.namespaces.stream().anyMatch(ns -> hook.startsWith(ns + "."));
+    }
+
+    public static final Production INJ_PROD = Production(KLabel(KLabels.INJ, Sort("S1"), Sort("S2")), Sort("S2"), Seq(NonTerminal(Sort("S1"))), Att());
+
+    private Production production(KApply term) {
+        return production(term, false);
+    }
+    private Production production(KApply term, boolean instantiatePolySorts) {
+        KLabel klabel = term.klabel();
+        if (klabel.name().equals(KLabels.INJ))
+            return instantiatePolySorts ? INJ_PROD.substitute(term.klabel().params()) : INJ_PROD;
+        Option<scala.collection.Set<Production>> prods = module.productionsFor().get(klabel.head());
+        if (!(prods.nonEmpty() && prods.get().size() == 1))
+            throw KEMException.compilerError("Expected to find exactly one production for KLabel: " + klabel + " found: " + prods.getOrElse(Collections::Set).size());
+        return instantiatePolySorts ? prods.get().head().substitute(term.klabel().params()) : prods.get().head();
+    }
+
+    private boolean isGeneratedInKeysOp(Production prod) {
+        Option<String> hook = prod.att().getOption(Att.HOOK());
+        if (hook.isEmpty()) return false;
+        if (!hook.get().equals("MAP.in_keys")) return false;
+        return (!prod.klabel().isEmpty());
+    }
+
+    public synchronized Sentence add(Sentence s) {
+        if (!(s instanceof Production))
+            return s;
+
+        Production prod = (Production) s;
+
+        SetMultimap<KLabel, Rule> functionRules = HashMultimap.create();
+        for (Rule rule : iterable(module.sortedRules())) {
+            K left = RewriteToTop.toLeft(rule.body());
+
+            if (left instanceof KApply) {
+                KApply kapp = (KApply) left;
+                    Production prod2 = production(kapp);
+                    if (prod2.att().contains(Att.FUNCTION()) || rule.att().contains(Att.ANYWHERE())
+                            || ExpandMacros.isMacro(rule)) {
+                        functionRules.put(kapp.klabel(), rule);
+                    }
+            }
+        }
```

Comment:
This will be optimized in a future PR. The idea is to move this and the production's functions to the module definition and call it here and in the ModuleKore file

Reply:
Addressed in https://github.com/runtimeverification/k/pull/3522/commits/4e062801eea8b1d6fd165783f4fa7cb7bbc6b208